timely_beliefs/sensors/func_store/knowledge_horizons.py

```diff
@@ -31,6 +32,45 @@ def at_date(
     return event_start - knowledge_time.astimezone(event_start.tzinfo)
 
 
+
+def x_years_ago_at_date(
+    event_start: datetime | pd.DatetimeIndex,
+    day: int,
+    month : int,
+    x : int = 1,
+    get_bounds: bool = False,
+    z: str | None = None,
+) -> timedelta | pd.TimedeltaIndex | tuple[timedelta, timedelta]:
+    """Compute the sensor's knowledge horizon to represent the event could be known since some date, `x` years ago.
+    
+    For example, it can be used for a tax rate that changes annually and with a known publication date. 
+
+    :param event_start:     Start of the event, used as an anchor for determining the knowledge horizon.
+    :param day:             Reference day of the month of the annual date to compare against.
+    :param month:           The month of the annual date to compare against.
+    :param x:               The number of years to shift the reference date to.
+    :param z:           Timezone string.
+    :param get_bounds:      If True, this function returns bounds on the possible return value.
+                            These bounds are normally useful for creating more efficient database queries when filtering by belief time.
+                            In this case, the knowledge horizon is unbounded.
+    """
+    if get_bounds:
+        return timedelta.min, timedelta.max
+    def x_years_ago_at_date_datetime(_event_start : datetime, day : int, month : int, x : int, z : str | None) -> timedelta:
+        if z is None:
+            z = _event_start.tzinfo
+        else:
+            z = timezone(z)
+
+        anchor = dict(year=_event_start.year - x, month=month, day=day, hour=0, minute=0, second=0, microsecond=0, tzinfo=z)
+
+        return _event_start - _event_start.replace(**anchor)
+        
+    if isinstance(event_start, datetime):
+        return x_years_ago_at_date_datetime(event_start, day, month, x, z)
+    else:
+        return event_start.map(lambda _event_start: x_years_ago_at_date_datetime(_event_start,day,month, x, z))
```

Comment:
It would help speed if we can add a vectorized implementation for the DatetimeIndex case: see https://github.com/SeitaBV/timely-beliefs/blob/38e285effb8fc8521dfe235a3ea3730bcffd2952/timely_beliefs/sensors/func_store/utils.py#L38 for an example.

Reply:
I made a try but there is no way to replace values of a DateTimeIndex with a vectorized method. Also, I tried to use a combination of timedelta and floor, but we can't apply those with a yearly frequency.

Reply:
What about something like `.to_period('Y').to_timestamp()`?

Reply:
Thanks! I used that in combination with `DateOffset` to vectorize it.